content/blog/throttling-coredns.md

```diff
@@ -0,0 +1,141 @@
++++
+title = "Throttling CoreDNS"
+description = "A guide to limit impact on memory consumotion due to burst of queries"
+tags = ["Kubernetes", "Memory", "DNS", "Throttle", "Documentation"]
+date = "2019-03-26T20:23:43-00:00"
+author = "sandeep"
++++
+
+## CoreDNS Memory Consumption
+
+There have been [some incidents reported](https://github.com/coredns/coredns/issues/2593)
+regarding CoreDNS taking up too much memory when there is a high rate of DNS queries.
+
+In the context of Kubernetes, this results in CoreDNS crashing as it gets `OOMkilled`.
+To understand this better, we need to understand the resource requests and limits of a pod and container.
+A deep dive on how Kubernetes manages its resources for containers is well documented [here](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/).
+To give a high level summary, when CoreDNS exceeds its resource limits and requests the following may occur:
+ - If the CoreDNS pod exceeds its memory limit, it might be terminated. If it is restartable, the kubelet will restart it.
+ - If CoreDNS exceeds its memory request, it is likely that its Pod will be evicted whenever the node runs out of memory.
+ 
+Below is the CoreDNS limits set in Kubernetes:
+
+~~~ yaml
+resources:
+  limits:
+    memory: 170Mi
+  requests:
+    cpu: 100m
+    memory: 70Mi
+~~~
+
+This means that if CoreDNS uses greater than 170Mi due to a burst of incoming DNS queries, CoreDNS will be `OOMKilled` due to it taking up more
+memory than the limit set. This may be problematic if the CoreDNS pods are crashing since it may lead to an outage.
+
+To prevent the memory spike from happening, the `throttle` plugin has been proposed.
+The proposed `throttle` plugin helps to limit the maximum number of simultaneous inflight queries which helps in preventing CoreDNS
+being `OOMkilled` in case of a burst of incoming queries. 
+
+## Testing CoreDNS memory usage
+
+To test the `throttle` plugin and show the difference in memory consumption of CoreDNS with and without the `throttle` plugin,
+a Kubernetes cluster was set up with one replica of CoreDNS running in the Master node. The memory limit of the CoreDNS container was set
+to 1700Mi to help capture the memory consumption in case CoreDNS crashes due to it being `OOMkilled`.
+
+To provide a burst of incoming queries, [a simple tool for stressing CoreDNS](https://github.com/mikkeloscar/go-dnsperf) was used.
```

Comment:
Probably should explain why this tool, and not an existing more widely accepted tool such as `dnsperf`.